lib/db/leveldb_dbinstance.go

```diff
@@ -186,20 +186,27 @@ func (db *Instance) removeSequences(folder []byte, fs []protocol.FileInfo) {
 }
 
 func (db *Instance) withHave(folder, device, prefix []byte, truncate bool, fn Iterator) {
+	if len(prefix) > 0 && !bytes.HasSuffix(prefix, []byte{'/'}) {
+		f, ok := db.getFileTrunc(db.deviceKey(folder, device, prefix), true)
+		if !ok {
+			// nothing to be done if the root element doesn't exist
+			return
+		}
+		if !fn(f) {
+			return
+		}
+		prefix = append(prefix, '/')
+	}
+
 	t := db.newReadOnlyTransaction()
 	defer t.close()
 
 	dbi := t.NewIterator(util.BytesPrefix(db.deviceKey(folder, device, prefix)[:keyPrefixLen+keyFolderLen+keyDeviceLen+len(prefix)]), nil)
 	defer dbi.Release()
 
-	slashedPrefix := prefix
-	if !bytes.HasSuffix(prefix, []byte{'/'}) {
-		slashedPrefix = append(slashedPrefix, '/')
-	}
-
 	for dbi.Next() {
 		name := db.deviceKeyName(dbi.Key())
-		if len(prefix) > 0 && !bytes.Equal(name, prefix) && !bytes.HasPrefix(name, slashedPrefix) {
+		if len(prefix) > 0 && !bytes.HasPrefix(name, prefix) {
 			return
```

Comment:
So this no longer walks the directory itself? I think that's not how it should work, as I think walk returns the directory.

Reply:
I think this handled at the top.